src/interpreter/mod.rs

```diff
@@ -0,0 +1,1320 @@
+                call.ident(),
+                expected_ty,
+                actual_ty,
+                line,
+            ),
+        }
+    }
+}
+
+impl error::Error for RuntimeError {}
+
+/// An interpreter error.
+#[derive(Debug, PartialEq)]
+pub enum InterpretError {
+    Resolve(ResolveError),
+    Typecheck(TypecheckError),
+    Runtime(RuntimeError),
+}
+
+impl fmt::Display for InterpretError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            InterpretError::Resolve(resolve_error) => f.write_str(&resolve_error.to_string()),
+            InterpretError::Typecheck(typecheck_error) => f.write_str(&typecheck_error.to_string()),
+            InterpretError::Runtime(runtime_error) => f.write_str(&runtime_error.to_string()),
+        }
+    }
+}
+
+impl error::Error for InterpretError {}
+
+impl From<ResolveError> for InterpretError {
+    fn from(resolve_error: ResolveError) -> InterpretError {
+        InterpretError::Resolve(resolve_error)
+    }
+}
+
+impl From<TypecheckError> for InterpretError {
+    fn from(typecheck_error: TypecheckError) -> InterpretError {
+        InterpretError::Typecheck(typecheck_error)
+    }
+}
+
+impl From<RuntimeError> for InterpretError {
+    fn from(runtime_error: RuntimeError) -> InterpretError {
+        InterpretError::Runtime(runtime_error)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct VarInfo {
+    /// The parameters and function call this variable was created
+    /// with. Used to verify validity of this variable.
+    created_call: ast::CallExpr,
+
+    /// Current value of this variable.
+    value: Value,
+}
+
+/// Interpreter of a list of statements.
+///
+/// Allows to edit program and re-execute, trying to preserve computed
+/// data where possible. Program changes can be incremental, which
+/// allows for context preservation and less needless work. The
+/// smaller part of the program is overwritten, the less work needs to
+/// be done.
+pub struct Interpreter {
+    prog: ast::Prog,
+    funcs: Vec<Box<dyn Func>>,
+    env: HashMap<ast::VarIdent, VarInfo>,
+
+    /// The program counter. Always points to the **next** stmt to execute.
+    pc: usize,
+}
+
+impl Interpreter {
+    pub fn new(funcs: Vec<Box<dyn Func>>) -> Self {
+        Self {
+            prog: ast::Prog::default(),
+            funcs,
+            env: HashMap::new(),
+            pc: 0,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn prog(&self) -> &ast::Prog {
+        &self.prog
+    }
+
+    pub fn set_prog(&mut self, prog: ast::Prog) {
+        self.env.clear();
+        self.pc = 0;
+        self.prog = prog;
+    }
+
+    pub fn clear_prog(&mut self) {
+        self.env.clear();
+        self.pc = 0;
+        self.prog = ast::Prog::default();
+    }
+
+    pub fn push_prog_stmt(&mut self, stmt: ast::Stmt) {
+        self.prog.push_stmt(stmt);
+    }
+
+    pub fn pop_prog_stmt(&mut self) {
```

Comment:
True, but depends on whether we view popping the last statement off of an empty program as an error, or not. In our case the one calling this should know whether this will succeed beforehand. I initially thought it would be better not to hide this logic error by preventing a crash, but do not feel strongly about it - if you insist, I'll change it.